### DISCUSSION TITLE
(What-is) The best way to get the last line of a paragraph

### DISCUSSION THREAD

**Discussion by PhilippePerret** [prawnpdf/prawn#1316] [Q&A] [answered]
Hello,

I would like to know the best way to retrieve the last line of any given paragraph, in order to access its text and length before writing it to the PDF document (for the purpose of handling les "lignes de voleurs" — I don't know the english term, but it's not widow or orphan).

Thanks for any help, and sorry for bad english.

Phil

Phil

**Comment by pointlessone** [comment, marked as answer]
The closes thing to what you want that Prawn has is probably [`Prawn::Text::Formatted::Arranger`](https://github.com/prawnpdf/prawn/blob/8ceaa105f3e6038fac2eb4b84355d71b00d895f3/lib/prawn/text/formatted/arranger.rb) but keep in mind that it's not a public API and is not designed to be ergonomic.

**Reply by PhilippePerret** [reply]
Thanks Alexander, it seems perfect for my needs.

**Comment by PhilippePerret** [comment]
Maybe a best/easier way?

~~~ruby

my_string    = "bla bla bla" * 50
my_options = {
  font: ..., at: ...., width: ...,
  inline_format: true, 
  dry_run: true
}

pdf.update do

  exceeded_text, my_box = text_box(my_string, **my_options)
  # @note
  #   I overwrote #text_box method to deal with :dry_run
  #   (more convenient than box.render, see below)

  last_printed_line = my_box.instance_variable_get('@printed_lines').last
  # => last line text content

end
~~~

The overwritten #text_box method for my needs:

~~~ruby
  def text_box(string, options = {})
    options = options.dup
    options[:document] = self

    # @Added
    dry_run = options.delete(:dry_run)

    box =
      if options[:inline_format]
        p = options.delete(:inline_format)
        p = [] unless p.is_a?(Array)
        array = text_formatter.format(string, *p)
        Text::Formatted::Box.new(array, options)
      else
        Text::Box.new(string, options)
      end

    if dry_run
      exceed = box.render(dry_run: dry_run)
      [exceed, box]
    else
      [box.render, nil]
    end
  end

~~~

**Reply by pointlessone** [reply]
Whatever works for you. Either way it's unsupported by Prawn.
